subsys/bluetooth/mesh/access.c

```diff
@@ -2443,187 +2608,236 @@ static int read_comp_cb(const char *key, size_t len, settings_read_cb read_cb,
+	if (read_buf.len == 1 && read_buf.data[0] == 0) {
+		/* Single 0 byte is a sentinel value for empty page, return
+		 * success without writing any bytes to the buffer.
+		 */
+		return 0;
 	}
 
-	if (data_len) {
-		err = settings_save_one(path, data, data_len);
-	} else {
-		err = settings_delete(path);
+	if (offset > read_buf.len) {
+		return 0;
 	}
 
-	if (err) {
-		LOG_ERR("Failed to store %s value", path);
-	} else {
-		LOG_DBG("Stored %s value", path);
+	if (!allow_partial_elems) {
+		return write_cdp_elems(buf, &read_buf, page);
 	}
-	return err;
+
+	len = MIN(net_buf_simple_tailroom(buf), read_buf.len - offset);
+	net_buf_simple_add_mem(buf, read_buf.data + offset, len);
+
+	return 0;
 }
 
-#if defined(CONFIG_BT_MESH_LARGE_COMP_DATA_SRV)
-static int metadata_set(const char *name, size_t len_rd, settings_read_cb read_cb, void *cb_arg)
+static int stored_page_size_cb(const char *key, size_t len, settings_read_cb read_cb, void *cb_arg,
+			       void *param)
 {
-	/* Only need to know that the entry exists. Will load the contents on
-	 * demand.
-	 */
-	if (len_rd > 0) {
-		atomic_set_bit(bt_mesh.flags, BT_MESH_METADATA_DIRTY);
+	size_t *size = param;
+
+	if (len > 0) {
+		*size = len;
 	}
 
 	return 0;
 }
-BT_MESH_SETTINGS_DEFINE(metadata, "metadata", metadata_set);
 
-int bt_mesh_models_metadata_store(void)
+static size_t stored_page_size_get(enum page_type type, uint8_t page)
 {
-	NET_BUF_SIMPLE_DEFINE(buf, CONFIG_BT_MESH_MODELS_METADATA_PAGE_LEN);
-	size_t total_size;
 	int err;
+	size_t size = 0;
+	const char *path;
 
-	total_size = bt_mesh_metadata_page_0_size();
-	LOG_DBG("bt/mesh/metadata total %d", total_size);
-
-	net_buf_simple_init(&buf, 0);
-	net_buf_simple_add_le16(&buf, total_size);
+	path = stored_page_path(type, page);
 
-	err = bt_mesh_metadata_get_page_0(&buf, 0);
-	if (err == -E2BIG) {
-		LOG_ERR("Metadata too large");
-		return err;
-	}
-	if (err) {
-		LOG_ERR("Failed to read models metadata: %d", err);
-		return err;
+	if (path == NULL) {
+		return 0;
 	}
 
-	LOG_DBG("bt/mesh/metadata len %d", buf.len);
-
-	err = settings_save_one("bt/mesh/metadata", buf.data, buf.len);
+	err = settings_load_subtree_direct(path, stored_page_size_cb, &size);
 	if (err) {
-		LOG_ERR("Failed to store models metadata: %d", err);
-	} else {
-		LOG_DBG("Stored models metadata");
+		LOG_ERR("Failed getting stored page size for %sdata page %d: %d",
+			type == PAGE_TYPE_COMP ? "comp " : "meta", page, err);
+		return 0;
 	}
 
-	return err;
+	return size;
 }
+#endif /* CONFIG_BT_MESH_HIGH_DATA_PAGES */
 
-int bt_mesh_models_metadata_read(struct net_buf_simple *buf, size_t offset)
+static size_t page_size_get(enum page_type type, uint8_t page)
 {
-	NET_BUF_SIMPLE_DEFINE(stored_buf, CONFIG_BT_MESH_MODELS_METADATA_PAGE_LEN);
-	size_t original_len = buf->len;
-	int err;
+#ifdef CONFIG_BT_MESH_HIGH_DATA_PAGES
+	size_t size;
 
-	if (!IS_ENABLED(CONFIG_BT_SETTINGS)) {
-		return -ENOTSUP;
+	if (page >= 128) {
+		size = stored_page_size_get(type, page);
```

Comment:
Yes, this is by design. Basically, if there is no CDP stored in flash (meaning size == 0 here), that means that i.e. CDP129 == CDP1, so then size of CDP129 == current size of CDP1 (which is why I have the % 128 below). (Same for all pairs of data pages). I consider "no page in flash" to mean "same contents as current page" to avoid storing a copy of the low pages at all times, it is only stored in settings if it is different.